universe/supplyverifier/verifier.go

```diff
@@ -0,0 +1,641 @@
+	// step ensures that we have already verified the previous
+	// commitment, and that it is present in the database.
+	spentOutPoint, err := commitment.SpentCommitment.UnwrapOrErr(
+		fmt.Errorf("missing spent supply commitment outpoint"),
+	)
+	if err != nil {
+		return err
+	}
+
+	spentCommit, err :=
+		v.cfg.SupplyCommitView.FetchCommitmentByOutpoint(
+			ctx, assetSpec, spentOutPoint,
+		)
+	if err != nil {
+		return ErrPrevCommitmentNotFound
+	}
+
+	// Check that the given commitment spends the previous commitment's
+	// outpoint that is referenced by the given spent outpoint field.
+	checkSpendPrevOutPoint := false
+	for idx := range commitment.Txn.TxIn {
+		txIn := commitment.Txn.TxIn[idx]
+		if txIn.PreviousOutPoint == spentOutPoint {
+			checkSpendPrevOutPoint = true
+			break
+		}
+	}
+
+	if !checkSpendPrevOutPoint {
+		return fmt.Errorf("supply commitment does not spend " +
+			"provided previous commitment outpoint")
+	}
+
+	// Verify that every unspent pre-commitment output eligible by block
+	// height is actually spent by the supply commitment transaction.
+	err = v.ensurePrecommitsSpent(ctx, assetSpec, commitment)
+	if err != nil {
+		return fmt.Errorf("unable to verify pre-commitment spends: %w",
+			err)
+	}
+
+	// Get latest supply root tree and subtrees from the local db. Ensure
+	// that they correspond to the spent supply commitment outpoint.
+	spentRootTree, spentSubtrees, err :=
+		v.cfg.SupplyTreeView.FetchSupplyTrees(
+			ctx, assetSpec,
+		)
+	if err != nil {
+		return fmt.Errorf("unable to fetch spent root supply tree: %w",
+			err)
+	}
+
+	storedSpentRoot, err := spentRootTree.Root(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to compute root of local spent "+
+			"supply tree: %w", err)
+	}
+
+	if storedSpentRoot.NodeHash() != spentCommit.SupplyRoot.NodeHash() {
+		return fmt.Errorf("local spent supply tree root does not " +
+			"match spent commitment supply root")
+	}
+
+	// Apply new leaves to the spent subtrees to generate the new set of
+	// supply subtrees.
+	newSupplyTrees, err := supplycommit.ApplyTreeUpdates(
+		*spentSubtrees, leaves.AllUpdates(),
+	)
+	if err != nil {
+		return fmt.Errorf("unable to apply tree updates to spent "+
+			"commitment: %w", err)
+	}
+
+	// Reconstruct the root supply tree by applying the new leaves to
+	// the previous root supply tree.
+	expectedSupplyTree, err := supplycommit.UpdateRootSupplyTree(
+		ctx, spentRootTree, newSupplyTrees,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to generate expected root supply "+
+			"tree: %w", err)
+	}
+
+	expectedRoot, err := expectedSupplyTree.Root(ctx)
+	if err != nil {
+		return fmt.Errorf("unable to compute root of expected supply "+
+			"tree: %w", err)
+	}
+
+	// Ensure that the root of the reconstructed supply tree matches
+	// the commitment root.
+	if expectedRoot.NodeHash() != commitment.SupplyRoot.NodeHash() {
+		return fmt.Errorf("expected supply tree root does not match " +
+			"commitment supply root")
+	}
+
```

Comment:
Final missing thing here is to verify that this is actually what's committed to on chain (using the delegation key and the commitment we created arrives back at the same output). 

We should also verify that the pre commitments are using the delegation key as well.